src/functional/ffront/foast_passes/closure_var_folding.py

```diff
@@ -0,0 +1,79 @@
+# GT4Py Project - GridTools Framework
+#
+# Copyright (c) 2014-2022, ETH Zurich
+# All rights reserved.
+#
+# This file is part of the GT4Py project and the GridTools framework.
+# GT4Py is free software: you can redistribute it and/or modify it under
+# the terms of the GNU General Public License as published by the
+# Free Software Foundation, either vervisit_Constantsion 3 of the License, or any later
+# version. See the LICENSE.txt file at the top-level directory of this
+# distribution for a copy of the license or check <https://www.gnu.org/licenses/>.
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+from typing import Any
+
+import functional.ffront.field_operator_ast as foast
+from eve import NodeTranslator, traits
+from eve.utils import FrozenNamespace
+
+
+class ClosureVarFolding(NodeTranslator, traits.VisitorWithSymbolTableTrait):
+    """
+    Replace `Name` nodes that refer to closure variable by `Constant` nodes.
+
+    `Attribute` nodes are also replaced by a `Constant` nodes with the value of the
+     closure variable's corresponding attribute. This is executed recursively for
+     attributes of attributes.
+    """
+
+    closure_vars: dict[str, Any]
+    _current_closure_vars: list[foast.Symbol]
+
+    def __init__(self, closure_vars: dict[str, Any]):
+        self.closure_vars = closure_vars
+
+    @classmethod
+    def apply(cls, node: foast.FieldOperator, closure_vars: dict[str, Any]):
+        return cls(closure_vars=closure_vars).visit(node)
+
+    def visit_Name(self, node: foast.Name, **kwargs):
+        symtable = kwargs["symtable"]
+        if node.id in symtable:
+            definition = symtable[node.id]
+            if definition in self._current_closure_vars:
+                value = self.closure_vars[node.id]
+                if isinstance(value, FrozenNamespace):
+                    return foast.Constant(value=value, location=node.location)
```

Comment:
```suggestion
        if node.id in self.closure_vars and self.closure_vars[node.id] == symtable[node.id]:
            return foast.Constant(value=value, location=node.location)
```
If it is something we don't understand the type deduction will fail, no need to special case here.

Reply:
Without special casing, it would also convert callables and types into `Constant` nodes.